infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -91,33 +160,105 @@ std::vector<std::byte> read_memory(pid_t pid, unsigned long long address, size_t
       return memory;
     }
 
-    std::byte *word_bytes = reinterpret_cast<std::byte*>(&word);
-    memory.insert(memory.end(), word_bytes, word_bytes+sizeof(long));
+    std::byte *word_bytes = reinterpret_cast<std::byte *>(&word);
+    memory.insert(memory.end(), word_bytes, word_bytes + sizeof(long));
   }
 
   return memory;
 }
 
-void inspect(pid_t pid, const user_regs_struct &regs) {
-  auto memory = read_memory(pid, regs.rdi, kTripWire.length());
-  if (memory.size() == 0) {
-    return;
+// Construct a string with the memory specified in a register.
+std::string read_string(pid_t pid, unsigned long reg, unsigned long length) {
+  auto memory = read_memory(pid, reg, length);
+  if (!memory.size()) {
+    return "";
   }
 
-  std::string path(reinterpret_cast<char*>(
-        memory.data()), std::min(memory.size(), kTripWire.length()));
+  std::string content(reinterpret_cast<char *>(memory.data()),
+                      std::min(memory.size(), length));
+  return content;
+}
+
+void report_bug(std::string bug_type) {
+  // Report the bug found based on the bug code.
+  std::cerr << "===BUG DETECTED: " << bug_type.c_str() << "===\n";
+  // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
+  // to the root process.
+  // Note: this may not be reliable or consistent if shell injection happens
+  // in an async way.
+  kill(g_root_pid, SIGABRT);
+  _exit(0);
+}
+
+void inspect_for_injection(pid_t pid, const user_regs_struct &regs) {
+  // Inspect a PID's registers for the sign of shell injection.
+  std::string path = read_string(pid, regs.rdi, kTripWire.length());
+  if (!path.length()) {
+    return;
+  }
   debug_log("inspecting");
   if (path == kTripWire) {
-    fprintf(stderr, "===BUG DETECTED: Shell injection===\n");
-    // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
-    // to the root process.
-    // Note: this may not be reliable or consistent if shell injection happens
-    // in an async way.
-    kill(g_root_pid, SIGABRT);
-    _exit(0);
+    report_bug(kInjectionError);
   }
 }
 
+std::string get_pathname(pid_t pid, const user_regs_struct &regs) {
+  // Parse the pathname from the memory specified in the RDI register.
+  std::string raw_content = read_string(pid, regs.rdi, kShellPathnameLength);
+
+  std::string pathname = raw_content.substr(0, raw_content.find(" "));
```

Comment:
why do we need to stop at the first space?

Reply:
Because `kShellPathnameLength` is often longer than needed, hence the `raw_content` is often `/usr/bin/sh XXX YYY`.
We stop at the first space to get rid of the `XXX YYY` and only keep the shell part.

Reply:
That seems like it shouldn't happen in practice. The first argument to execve should be the path to the binary and not include any arguments.

See https://man7.org/linux/man-pages/man2/execve.2.html

Reply:
Just tested it again: You are right.
Fixed this and [the issue with `.find("\n")`](https://github.com/google/oss-fuzz/pull/7795#discussion_r888609245), not sure why I got those strange outputs earlier...
Thanks!

By the way,  I also addressed most of your other comments.
Once I fix the flaw in shell matching, I will push them together : )